genesis/engine/solvers/rigid/gjk_decomp.py

```diff
@@ -0,0 +1,3394 @@
+                # which requires the default GJK handling.
+                if distance > self.collision_eps:
+                    radius_a, radius_b = 0.0, 0.0
+                    if is_sphere_swept_geom_a:
+                        radius_a = self._solver.geoms_info[i_ga].data[0]
+                    if is_sphere_swept_geom_b:
+                        radius_b = self._solver.geoms_info[i_gb].data[0]
+
+                    wa = self.witness[i_b, 0].point_obj1
+                    wb = self.witness[i_b, 0].point_obj2
+                    n = self.func_safe_normalize(wb - wa)
+
+                    self.distance[i_b] = distance - (radius_a + radius_b)
+                    self.witness[i_b, 0].point_obj1 = wa + (radius_a * n)
+                    self.witness[i_b, 0].point_obj2 = wb - (radius_b * n)
+
+                    break
+
+            # Only try shrinking the sphere once
+            shrink_sphere = False
+
+        distance = self.distance[i_b]
+        nsimplex = self.gjk_nsimplex[i_b]
+        collided = distance < self.collision_eps
+
+        # To run EPA, we need following conditions:
+        # 1. We did not find min. distance with shrink_sphere flag
+        # 2. We have a valid GJK simplex (nsimplex > 0)
+        # 3. We have a collision (distance < collision_epsilon)
+        do_epa = (not shrink_sphere) and collided and (nsimplex > 0)
+
+        if do_epa:
+
+            # Assume touching
+            self.distance[i_b] = 0
+
+            # Initialize polytope
+            self.polytope[i_b].nverts = 0
+            self.polytope[i_b].nfaces = 0
+            self.polytope[i_b].nfaces_map = 0
+            self.polytope[i_b].horizon_nedges = 0
+
+            # Construct the initial polytope from the GJK simplex
+            polytope_flag = 0
+            if nsimplex == 2:
+                polytope_flag = self.func_epa_init_polytope_2d(i_ga, i_gb, i_b)
+            elif nsimplex == 4:
+                polytope_flag = self.func_epa_init_polytope_4d(i_ga, i_gb, i_b)
+
+            # Polytope 3D could be used as a fallback for 2D and 4D cases, but it is not necessary
+            if nsimplex == 3 or (polytope_flag == EPA_P2_FALLBACK3 or polytope_flag == EPA_P4_FALLBACK3):
+                polytope_flag = self.func_epa_init_polytope_3d(i_ga, i_gb, i_b)
+
+            # Run EPA from the polytope
+            if polytope_flag == 0:
+                i_f = self.func_epa(i_ga, i_gb, i_b)
+                distance = self.distance[i_b]
+
+                if num_max_contacts > 1 and i_f != -1:
+                    # Detect multiple contacts
+                    self.func_multi_contact(i_ga, i_gb, i_b, i_f)
+
+    """
+    GJK algorithms
+    """
+
+    @ti.func
+    def func_gjk(self, i_ga, i_gb, i_b, shrink_sphere):
+        """
+        GJK algorithm to compute the minimum distance between two convex objects.
+        """
+
+        # Simplex index
+        n = 0
+
```

Comment:
I would recommend not skipping space between top-level parameters to keep them grouped. It improves readability I think.